mesonbuild/dependencies/dub.py

```diff
@@ -223,33 +192,45 @@ def find_package_target(pkg: DubPackDesc) -> bool:
         # 4. Add other build settings (imports, versions etc.)
 
         # 1
-        self.is_found = False
         packages: T.Dict[str, DubPackDesc] = {}
+        found_it = False
         for pkg in description['packages']:
             packages[pkg['name']] = pkg
 
             if not pkg['active']:
                 continue
 
-            if pkg['targetType'] == 'dynamicLibrary':
-                mlog.error('DUB dynamic library dependencies are not supported.')
-                self.is_found = False
-                return
-
             # check that the main dependency is indeed a library
             if pkg['name'] == name:
-                self.is_found = True
-
                 if pkg['targetType'] not in ['library', 'sourceLibrary', 'staticLibrary']:
-                    mlog.error(mlog.bold(name), "found but it isn't a library")
-                    self.is_found = False
+                    mlog.error(mlog.bold(name), "found but it isn't a static library, it is:",
+                               pkg['targetType'])
                     return
 
+                if self.version_reqs is not None:
+                    ver = pkg['version']
+                    if not version_compare_many(ver, self.version_reqs)[0]:
+                        mlog.error(mlog.bold(f'{name}@{ver}'),
+                                   'does not satisfy all version requirements of:',
+                                   ' '.join(self.version_reqs))
+                        return
+
+                found_it = True
                 self.version = pkg['version']
                 self.pkg = pkg
 
+        if not found_it:
+            mlog.error(f'Could not find {name} in DUB description.')
+            if description_source is DubDescriptionSource.Local:
+                mlog.log('Make sure that the dependency is registered for your dub project by running:')
+                mlog.log(mlog.bold(f'dub add {name}'))
+            elif description_source is DubDescriptionSource.External:
+                # It shouldn't be possible to get here
```

Comment:
Should we assert or ask them to file a bug then?

Reply:
Changed this to raise an exception explaining why this code should be unreachable. It does depend on the behavior of `dub` so maybe it's safer to actually treat the error